lib/engine/game/g_1858_india/step/collect_tokens.rb

```diff
@@ -0,0 +1,139 @@
+            return [] unless entity.corporation?
+            return [] unless can_collect_token?(entity, false)
+
+            ACTIONS
+          end
+
+          def auto_actions(entity)
+            # Checking whether a corporation has a route to an available token
+            # involves recalculating the game graphs, which is expensive, so
+            # `actions` only checks if there is an available token somewhere on
+            # the board.
+            # To avoid asking the player to choose tokens when none are really
+            # available, insert an auto-action pass.
+            return unless entity.corporation?
+            return if can_collect_token?(entity, true)
+
+            [Engine::Action::Pass.new(entity)]
+          end
+
+          def description
+            'Collect mine, oil and port tokens'
+          end
+
+          def help
+            'Collect mine tokens ' \
+              "(#{@game.format_currency(TOKEN_COST['mine'])} each), " \
+              'oil tokens ' \
+              "(#{@game.format_currency(TOKEN_COST['oil'])} each) " \
+              'and port tokens ' \
+              "(#{@game.format_currency(TOKEN_COST['port'])} each)."
+          end
+
+          def log_skip(entity)
+            super unless entity.minor?
+          end
+
+          def available_hex(entity, hex)
+            can_remove_hex_token?(entity, hex)
+          end
+
+          def can_remove_hex_token?(entity, hex)
+            connected_tokens(entity).map(&:hex).include?(hex)
+          end
+
+          def process_remove_hex_token(action)
+            corporation = action.entity
+            hex = action.hex
+            token = hex.tokens.first
+            price = token_price(token)
+
+            msg = "#{corporation.id} "
+            msg += "spends #{@game.format_currency(price)} and " if price.positive?
+            msg += "collects a #{token_type(token)} token from " \
+                   "hex #{hex.coordinates}."
+            @game.log << msg
+            corporation.spend(price, @game.bank) if price.positive?
+            hex.remove_token(token)
+            corporation.tokens << token
+          end
+
+          private
+
+          def can_collect_token?(entity, check_graph)
+            if check_graph
+              !connected_tokens(entity).empty?
+            else
+              !available_tokens(entity).empty?
+            end
+          end
+
+          # Returns all mine/oil/port tokens on the map which have not yet
+          # been collected, where there has been track laid in the hex.
+          # For port tokens the corporation must be able to afford the token
+          # and the current phase must allow them to be taken.
+          def available_tokens(corporation)
+            tokens = []
+            tokens += mines if corporation.cash >= TOKEN_COST['mine']
+            tokens += oil if corporation.cash >= TOKEN_COST['oil']
+            tokens += ports if corporation.cash >= TOKEN_COST['port']
```

Comment:
A minor point, += creates a new array, whereas tokens.concat(oil) will add the elements of oil to the existing tokens array. It won't make much difference for very small arrays but is generally more efficient to use concat().

Reply:
OK. I had thought that `tokens.concat(oil)` was the same as `tokens << oil` which doesn't do what I need.